grails-app/services/org/pih/warehouse/core/WebhookPublisherService.groovy

```diff
@@ -0,0 +1,71 @@
+/**
+* Copyright (c) 2012 Partners In Health.  All rights reserved.
+* The use and distribution terms for this software are covered by the
+* Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
+* which can be found in the file epl-v10.html at the root of this distribution.
+* By using this software in any fashion, you are agreeing to be bound by
+* the terms of this license.
+* You must not remove this notice, or any other, from this software.
+**/
+package org.pih.warehouse.core
+
+import org.codehaus.groovy.grails.commons.ConfigurationHolder
+import org.pih.warehouse.shipping.Shipment
+import org.pih.warehouse.auth.AuthService
+
+
+class WebhookPublisherService {
+
+    def apiClientService
+    //def grailsApplication
+
+    boolean transactional = false
+
+    def publishShippedEvent(Shipment shipment) {
+
+        boolean webhooksEnabled = shipment.origin.supports(ActivityCode.ENABLE_WEBHOOKS)
+        if (!webhooksEnabled) {
+            log.info "Location ${shipment.origin} does not support activity code ${ActivityCode.ENABLE_WEBHOOKS}"
+            return
+        }
+
+        Map payload = [
```

Comment:
Perhaps it's worth moving into `Shipment`, as `toWebhookJson()` or something like that (?)

Reply:
Or this should be related to a specific webhook? What I imagine is there might be a couple of webhooks configured at the same time and each could have its own "mapping". So perhaps there would be a better place for webhook payload models instead of having this in the shipment?

Reply:
I was letting the implementation of this marinate in my head a bit. I agree each domain should know what it's payload is, but this isn't just a webhook thing. Instead it needs to support generic notifications about each domain and state change. We are planning to deal with this in the improved webhook feature in OBGM, so I think we can implement it there. You can review Chetan's PR #4300 and provide feedback there 

But essentially, I want the application to emit application-level and database-level events for everything that happens in the system (i.e. ShipmentStatusUpdateEvent, HibernateUpdateEvent, ProductCreatedEvent, UserLoginEvent, etc). 

A system administrator will be able to define different notification mechanisms: webhook, email, SMS, and push notifications. The application will allow the user to subscribe to any of the events that implement the NotifiableEvent interface (or extend it if we decide it should be an abstract class). For now, these subscriptions will be configured in the application configuration, but eventually, we'll have a table where we can manage the event notification subscriptions. 

When an application event is fired (we need to manually add these in the codebase for the app-level events, but we'll have a Hibernate persistence listener to publish the low-level database events) those events will be sent to the event bus. We'll have a listener listening for instances of NotifiableEvent or whatever we call them. These events need to provide enough information for us to send a notification, so we will convert them, I think, to a NotificationEvent which has a context (based on the source object) and configuration (type = webhook, SMS, email, templates, etc). The NotificationEvent is published to the same event bus but now the listeners are going to be implementations of a NotificationEventService which know how to publish to SMTP server, SMS endpoint, or webhook. The notification event service takes the notification context and the config to create the components needed for each type of notification i.e. email requires an string subject, HTML body, email recipients; webhook requires a URL, payload, headers. 

Anyway, that's the idea behind the improved version (#4300) but I still need to review the code to make sure we implemented it that way.

I think the latest discussion of the webhooks happened in the #moriana Slack channel, but it might be buried in a document or ticket somewhere as well.
https://openboxes.slack.com/archives/C05SDV26YKC/p1695918514682159

Here's the original ticket
https://pihemr.atlassian.net/browse/OBS-1515

cc @awalkowiak @drodzewicz @alannadolny @kkaczmarczyk @chetanmaharishi